src/compute/src/render/mod.rs

```diff
@@ -919,3 +920,57 @@ impl<T: Timestamp + Lattice> RenderTimestamp for Product<mz_repr::Timestamp, T>
         Product::new(self.outer.saturating_sub(1), self.inner.clone())
     }
 }
+
+/// Used to make possibly-validating code generic: think of this as a kind of `MaybeResult`,
+/// specialized for use in compute.  The contract is that the error constructor for non-validating
+/// implementors will never be called.
+trait MaybeValidatingRow<T, E>:
+    differential_dataflow::ExchangeData + columnation::Columnation + std::hash::Hash
```

Comment:
Nit: It would be appropriate to `use` these traits directly, rather than their modules. We do so throughout the codebase and their names are unambiguous.